java/src/main/java/ai/rapids/cudf/ColumnVector.java

```diff
@@ -265,32 +354,81 @@ public HostColumnVector copyToHost() {
       HostMemoryBuffer hostDataBuffer = null;
```

Comment:
Can we please make `copyToHost` call a recursive version of itself.  This is getting really confusing, and conceptually it should be very simple. I used `HostColumnVectorStandIn` here because conceptually the `copyToDevice` code should look very similar and I just wanted to show conceptually what it would look like. Here we might be able to get away with just a modified `HostColumnVector` or a `NestedHostColumnVector`.

```
private static final HostColumnVectorStandIn copyToHost(long deviceColumnView) {
  // Now some of this we could speed up with caching, but conceptually it should be like the following...
  // None of these need to be closed because they do not own any data, just views conceptually.
  DeviceMemoryBufferView offset = getOffsetView(deviceColumnView);
  DeviceMemoryBufferView valid = getValidView(deviceColumnView);
  DeviceMemoryBufferView data = getDataView(deviceColumnView);
  HostMemoryBuffer hostOffset = null;
  HostMemoryBuffer hostValid = null;
  HostMemoryBuffer hostData = null;
  List<HostColmnVectorStandIn> children = new List<>();
  try {
    for (long childView: getChildren(deviceColumnView)) { // Don't include offset children, etc
      children.add(copyToHost(childView));
    }
    // copy offset, valid and data to host as needed
    // Ownership of children and host buffers is transferred to new HostColumnVectorStandIn
    return new HostColumnVectorStandIn(hostOffset, hostValid, hostData, children);
  } finally {
    // close children if needed
    // close host buffers if needed
  }
}

public HostColumnVector copyToHost() {
  return copyToHost(getColumnView()).asHostColumnVector();
}
```

Reply:
>Can we please make copyToHost call a recursive version of itself.
>
Yes that is my intent, however since the previous version was all iterative, I am doing the rework onepiece at a time. There are a couple more places where I rely on the older lists of offsets, validity etc. WIll fix in the next round of commits.

Reply:
I have made `copytToHost` and `copyTodevice` recursive and also removed getBuffersToClose on the ColumnVector specifically. The recursive changes to these methods are close to what u imagined I think @revans2 but I could use some more input on this if u feel we need to further tighten the ownership when we do `createNestedColumnVector` and such.